p2p/src/network/rpc/p2p_network_rpc_reducer.rs

```diff
@@ -48,53 +98,233 @@ impl P2pNetworkRpcState {
+                        stream_id: *stream_id,
+                        message,
+                    })
                 }
+
+                Ok(())
             }
-            P2pNetworkRpcAction::IncomingMessage { message, .. } => {
+            action @ P2pNetworkRpcAction::IncomingMessage {
+                message,
+                addr,
+                peer_id,
+                stream_id,
+            } => {
                 if let RpcMessage::Response { header, .. } = message {
-                    if let Some(QueryHeader { id, tag, version }) = &self.pending {
-                        *self.total_stats.entry((tag.clone(), *version)).or_default() += 1;
+                    if let Some(QueryHeader { id, tag, version }) = &rpc_state.pending {
+                        *rpc_state
+                            .total_stats
+                            .entry((tag.clone(), *version))
+                            .or_default() += 1;
                         if id != &header.id {
-                            openmina_core::error!(action.time(); "receiving response with wrong id: {}", header.id);
+                            openmina_core::error!(meta.time(); "receiving response with wrong id: {}", header.id);
                         }
                     } else {
-                        openmina_core::error!(action.time(); "receiving response without query");
+                        openmina_core::error!(meta.time(); "receiving response without query");
                     }
                 } else if let RpcMessage::Query { header, .. } = message {
-                    if self.pending.is_none() {
-                        self.pending = Some(header.clone());
+                    if rpc_state.pending.is_none() {
+                        rpc_state.pending = Some(header.clone());
                     } else {
-                        openmina_core::error!(action.time(); "receiving query while another query is pending");
+                        openmina_core::error!(meta.time(); "receiving query while another query is pending");
+                    }
+                }
+
+                rpc_state.incoming.pop_front();
+
+                let (dispatcher, state) = state_context.into_dispatcher_and_state();
+                let network_state: &P2pNetworkState = state.substate()?;
+                let state = network_state
+                    .find_rpc_state(action)
+                    .ok_or_else(|| format!("RPC state not found for action: {:?}", action))?;
+
+                match &message {
+                    RpcMessage::Handshake => {
+                        if !state.is_incoming {
+                            dispatcher.push(P2pChannelsRpcAction::Ready { peer_id: *peer_id });
+                        }
+                    }
+                    RpcMessage::Heartbeat => {}
+                    RpcMessage::Query { header, bytes } => {
+                        if let Err(e) = rpc_query_effects(*peer_id, header, bytes, dispatcher) {
```

Comment:
It could be a bit confusing to have a function named `rpc_query_effects` called by the reducer

Reply:
true, when it comes to logic that just makes more dispatches we should move away from the "effects" terminology and reserve that for interactions with the "external world" (services).

@dkuehr any suggestion for a better naming?

Reply:
I'm not really sure, I remember there was some naming conventions in redux for these cases (an action that produces more actions) but I think we should stay away from it.